### DISCUSSION TITLE
fastapi app running in docker continuously increasing computer memory usage even when restarted

### DISCUSSION THREAD

**Discussion by kvalli** [fastapi/fastapi#12268] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
import fastapi

app = FastAPI()

uvicorn.run(app, host='localhost', port=8000)
```

### Description

I have two fastapi applications,both of which launch with one single "docker compose up" command.They start running .One App receives requests from unity.
The requests coming from unity come every 0.1 sec and the requests come until 3-4 minutes of unity game running.
after unity game ends,it restarts the game and fastapi app's are restarted.Since fastapi apps are restarted,even then the memory usage of app's is still around 1gb of virtual memory and overall usage of this docker application is not reduced,How to ensure that while the docker runs ,the memory usage does not overshoot.how to cleanup variables being used in endpoints,I'm also using lifespan to startup and shutdown to clean memory but this does not seem to help and no visible improvement 

### Operating System

Windows

### Operating System Details

windows11

### FastAPI Version

0.110.1

### Pydantic Version

2.6.4

### Python Version

3.12.2

**Comment by Cycloctane** [comment]
Do you really need to keep the containers running without restarting? If not, setting up restart policy and restarting containers along with your fastapi apps would be a better choice. See https://docs.docker.com/engine/containers/start-containers-automatically/#use-a-restart-policy

**Comment by hammad93** [comment]
Hello, I had this issue of a memory leak and was able to recently solve this, and it was running on a Docker container on top of the Anaconda 3 base image. Although I didn't run experiments with data to isolate the problem, these changes summarize the fix,

- Only use 1 worker at runtime
- Do not reload on change at runtime
- Remove global database clients, I had a Redis client as a global but instead reinitialized the client every time an endpoint is called

While I also implemented manual garbage collection by calling gc.collect(), this didn't help much but could also help.

**Comment by YuriiMotov** [comment]
Let's close this for now to keep the discussions clean.
If you’ve found a solution, please share it here.
If you still need help, feel free to let us know.
